data_pipeline/elife_article_xml/elife_article_xml_pipeline.py

```diff
@@ -0,0 +1,141 @@
+import requests
+
+
+from data_pipeline.elife_article_xml.elife_article_xml_config import (
+    ElifeArticleXmlConfig,
+    ElifeArticleXmlSourceConfig
+)
+from data_pipeline.utils.data_store.bq_data_service import (
+    does_bigquery_table_exist,
+    get_single_column_value_list_from_bq_query,
+    load_given_json_list_data_from_tempdir_to_bq
+)
+from data_pipeline.utils.json import (
+    get_recursively_transformed_object,
+    remove_key_with_null_value
+)
+from data_pipeline.utils.xml import parse_xml_and_return_it_as_dict
+
+LOGGER = logging.getLogger(__name__)
+
+
+def get_json_response_from_url(url: str) -> Any:
+    response = requests.get(url=url)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_url_of_xml_file_directory_from_repo(
+    source_config: ElifeArticleXmlSourceConfig,
+) -> str:
+    response_json = get_json_response_from_url(url=source_config.git_repo_url)
+    for folder in response_json['tree']:
+        if folder['path'] == source_config.directory_name:
+            return folder['url']
+
+
+def iter_unprocessed_xml_file_url_from_git_directory(
+    source_config: ElifeArticleXmlSourceConfig,
+    processed_file_url_list: Iterable[str]
+) -> Iterable[str]:
+    response_json = get_json_response_from_url(
+        url=get_url_of_xml_file_directory_from_repo(source_config=source_config)
+    )
+    for article_xml_url in response_json['tree']:
+        if article_xml_url['size'] > 0:
+            if article_xml_url['url'] not in processed_file_url_list:
+                yield article_xml_url['url']
+
+
+def iter_decoded_xml_file_content(
+    article_xml_url_list: Iterable[str]
+) -> Iterable[str]:
+    for article_xml_url in article_xml_url_list:
+        response_json = get_json_response_from_url(url=article_xml_url)
+        assert response_json['encoding'] == 'base64'
+        yield base64.b64decode(response_json['content']).decode('utf-8')
+
+
+def get_bq_compatible_transformed_key_value(
+    key: str,
+    value: Any
+) -> Tuple[Optional[str], Optional[Any]]:
+    return (
+        key.replace('-', '_'),
+        value
+    )
+
+
+def get_bq_compatible_json_dict(json_dict: dict) -> dict:
+    return get_recursively_transformed_object(
+        remove_key_with_null_value(json_dict),
+        key_value_transform_fn=get_bq_compatible_transformed_key_value
+    )
+
+
+def get_article_json_data_from_xml_string_content(
+    xml_string: str
+) -> dict:
+    xml_root = ET.fromstring(xml_string)
+    parsed_dict = parse_xml_and_return_it_as_dict(xml_root)
+    parsed_dict = get_bq_compatible_json_dict(parsed_dict)
+    if parsed_dict:
+        article_meta_dict = parsed_dict['article']['front'][0]['article_meta'][0]
+        for key in article_meta_dict.copy().keys():
+            if key != 'related_article':
+                article_meta_dict.pop(key, None)
+        if 'related_article' in article_meta_dict:
+            for key in article_meta_dict['related_article'][0].copy().keys():
```

Comment:
The accessing with zero look a bit suspicious. We could probably come up with a theoretical issue for `front` and `article-meta`. Although in practise it is probably fine. But here I would think it would cause an issue. You could of course just iterate over the items, e.g.

```python
for related_article in article_meta_dict.get('related_article', []):
    ...
```

To change remove the namespace though, you could also just modify `get_bq_compatible_transformed_key_value` as that is its main purpose. Either like you have done here:

```python
if 'href' in key:
    return 'href', value
```

Or a regular expression.

```python
key = re.sub(r'^{[^}]*}', '', key)
...
```